### ISSUE TITLE
Give IR visitor a return type.

### ISSUE THREAD

**Issue by aferr** [google-research/raksha#400]
That way we don't have to implement visitors just by side-effecting into the visitor data structure.

cc: @bgogul

**Comment by Cypher1** [comment]
Status update and some thoughts about the visitor pattern as we have it for Raksha's IR:

With our default `IRVisitor`, there's no traversal defined. Really the visitor there doesn't *do*, so the return values can't be used easily.

With the `IRTraversingVisitor`, we have almost the opposite problem. There's a fair bit of structure to change to get the results back without using side effecting/mutability (at which point it's not clear of the benefits).

I suggest that we update the IRTraversingVisitor:

a) Introduce a `Result` type parameter. (This would also be added to the `IRVisitor` to match and would be returned from each `Apply` and `Visit` function).
b) Introduce `ApplyResult` functions called with the result of each 'child' traversal, allowing results to be accumulated.
c) Default the `Result` type parameter to `std::monostate` (as `void` and `void&` cannot be constructed or passed as arguments).

```
void ApplyResult(const Module& module, Result child_result, Result& result);
void ApplyResult(const Block& block, Result child_result, Result& result);
void ApplyResult(const Operation& operation, Result child_result, Result& result);
```


This seems like a fairly opinionated change, and wouldn't be backwards compatible, though would be fairly trivial to implement.

With a single return type argument, most visitors will need each possible return value type to be combined in a `std::variant` or similar, and then need to do checking to make sure that the children are of the right type when taking the accumulated results.

This is all avoidable either:
1. By not using a visitor pattern, but that may lead to less maintainable code, as it's less clear how to ensure that changes in the AST & IR require changes in the code that traverses them.
2. By having more than one `Return` type parameter, which would allow us a more type safe, but perhaps less flexible implementation.

I'm currently writing the IR->Proto pass with approach 1. but I intended to finish this, test it and then approach converting it to use option 2. I'll post the results here for review.

Looking forward to comments.

**Comment by aferr** [comment]
Overall, I like your proposal. 

>I suggest that we update the IRTraversingVisitor:
> a) Introduce a Result type parameter. (This would also be added to the IRVisitor to match and would be > returned from each Apply and Visit function).
> b) Introduce ApplyResult functions called with the result of each 'child' traversal, allowing results to be accumulated.
> c) Default the Result type parameter to std::monostate (as void and void& cannot be constructed or
> passed as arguments).

Do you think it would be useful to also define a default _value_ in addition to the default return parameter that can be used for the `ApplyResult`s when the accumulation hasn't started yet. (Sort of like a functional `fold` with a default value.) For example, a class that extends from your new visitor might override this default value to specify that it's the empty set, or a set with one specific element, or so on.

> This seems like a fairly opinionated change, and wouldn't be backwards compatible, though would be fairly trivial to implement.

Can the lack of backwards compatibility be remedied by (approximately) leaving the IRTraversingVisitor as is and writing your new version as a class that extends from `IRVisitor` and mirrors the `IRTraversingVisitor` aside from the changes you proposed? (At least one change to the `IRTraversingVisitor` that would still be needed is the Result parameter, so at a glance this would still require some changes to `IRTraversingVisitor`'s API... so this might not work out).

If writing a class that's parallel to IRTraversingVisitor does not work, I think I don't mind the change to the code that consumes the `IRTraversingVisitor`.

> This is all avoidable either:

> 1. By not using a visitor pattern, but that may lead to less maintainable code, as it's less clear how to ensure that changes in the AST & IR require changes in the code that traverses them.
> 2. By having more than one Return type parameter, which would allow us a more type safe, but perhaps less flexible implementation.

Of these I prefer option 2. I think there are:
 - some use-cases for visitors where the application of the visitor winds up repeating the entire traversal (because what happens between visits is important), and so the visitor pattern is not helping
 - other use-cases where the traversal order does not matter and the visitor vastly improves things

Because there are at least a lot of cases where the visitor does help, it's useful to still use a visitor.

**Comment by bgogul** [comment]
Overall, the plan seems fine. 

> I'm currently writing the IR->Proto pass with approach 1.

You might still be able to use the IRTraversingVisitor by abusing the state in the visitor. e.g., `ir::proto::Module* parent_module_`, `ir::proto::Block* parent_block`, etc. However, if it turns awkward, let us not force it. 

>By having more than one Return type parameter, which would allow us a more type safe, but perhaps less flexible implementation.

@Cypher1 are you suggesting having more than one return type in the template argument? e.g., one for Module visit, one for Operation visit, etc. 

>Of these I prefer option 2. I think there are:
I think it would better to have one return type for to reduce cognitive load. The typical workflow is that we return a single type after a visit anyway? If there is need to return multiple return types, the user could fold it into a variant?

> b) Introduce ApplyResult functions called with the result of each 'child' traversal, allowing results to be accumulated.
May be call it `CombineResult` or `FoldResult`?

> c) Default the Result type parameter to std::monostate (as void and void& cannot be constructed or passed as arguments).
TIL

**Comment by Cypher1** [comment]
Thanks Gogul,

Yes, if I use the std::variant approach along with Combine or Fold Result, I should be fine to use the Traversing visitor **and** the visitors should all able to be made (mostly) stateless (things like tracking unique IDs might still be required to be side effects).

I'm pretty keen on this approach and I think it actually protects against future changes in the IR AST better than my 'multiple return type' approach, which I have discussed as an alternative (but do not plan to implement without feedback).

I am looking at the dispatcher you showed me though, and so I'll try to ensure that we appropriately enforce constness constraints so that the one traverser can update the IR as well as just inspect it.

I've started writing tests for the round tripping with my hand written conversions and then I should get it into visitor form early this coming week. I'll put up a draft once I get tests passing so I can get feedback on the style and C++ idioms.

**Comment by Cypher1** [comment]
Oh, and I think `FoldResult` is probably the better name, as it explains the way that the results will be applied (where Apply and Combine leave the approach unclear). Hopefully folding is common enough these days to be easily understandable.

**Comment by Cypher1** [comment]
Ah, sorry @aferr I got the emails out of order and didn't realise that I'd missed your comments.

> Do you think it would be useful to also define a default value in addition to the default return parameter that can be used for the ApplyResults when the accumulation hasn't started yet. (Sort of like a functional fold with a default value.) For example, a class that extends from your new visitor might override this default value to specify that it's the empty set, or a set with one specific element, or so on.

I think perhaps it would be better to have `Previsit` return the starting value. This could support a default value but could also do something based on the internal state, a little more flexibly.

> Can the lack of backwards compatibility be remedied by (approximately) leaving the IRTraversingVisitor as is and writing your new version as a class that extends from IRVisitor and mirrors the IRTraversingVisitor aside from the changes you proposed?
> (At least one change to the IRTraversingVisitor that would still be needed is the Result parameter, so at a glance this would still require some changes to IRTraversingVisitor's API... so this might not work out).

I think adding it 'beside' the current `IRTraversingVisitor` seems like a good plan. Then we can move things off using the old one if this is useful, without any changes. I'd prefer to either update or not update it though, adding a `Result` type param that isn't very useful... well it seems more confusing.

Thanks again for the feedback & suggestions :)

**Comment by Cypher1** [comment]
Draft PR at https://github.com/google-research/raksha/pull/598
I'm experimenting with some additional functions that make the change backwards compatible, while providing a little more flexibility for users w.r.t. return values.

I do want to check that it's not too complicated though.

**Comment by Cypher1** [comment]
Fixed in #598
